packages/@aws-cdk/toolkit-lib/lib/toolkit/toolkit.ts

```diff
@@ -249,6 +253,157 @@ export class Toolkit extends CloudAssemblySourceBuilder implements AsyncDisposab
+          options.compareAgainstProcessedTemplate,
+        );
+        const currentTemplate = templateWithNestedStacks.deployedRootTemplate;
+        const nestedStacks = templateWithNestedStacks.nestedStacks;
+
+        const migrator = new ResourceMigrator({ deployments, ioHelper });
+        const resourcesToImport = await migrator.tryGetResources(await deployments.resolveEnvironment(stack));
+        if (resourcesToImport) {
+          removeNonImportResources(stack);
+        }
+
+        let changeSet = undefined;
+
+        if (options.changeSet ?? true) {
+          let stackExists = false;
+          try {
+            stackExists = await deployments.stackExists({
+              stack,
+              deployName: stack.stackName,
+              tryLookupRole: true,
+            });
+          } catch (e: any) {
+            await ioHelper.notify(IO.CDK_TOOLKIT_I4201.msg(`Checking if the stack ${stack.stackName} exists before creating the changeset has failed, will base the diff on template differences.\n`));
+            await ioHelper.notify(IO.CDK_TOOLKIT_I4200.msg(formatErrorMessage(e)));
+            stackExists = false;
+          }
+
+          if (stackExists) {
+            changeSet = await createDiffChangeSet(ioHelper, {
+              stack,
+              uuid: uuid.v4(),
+              deployments,
+              willExecute: false,
+              sdkProvider: await this.sdkProvider('diff'),
+              parameters: Object.assign({}, parameterMap['*'], parameterMap[stack.stackName]),
+              resourcesToImport,
+            });
+          } else {
+            await ioHelper.notify(IO.CDK_TOOLKIT_I4300.msg(`the stack '${stack.stackName}' has not been deployed to CloudFormation or describeStacks call failed, skipping changeset creation.`));
+          }
+        }
+
+        if (options.securityOnly) {
+          const securityDiff = formatSecurityDiff(
+            ioHelper,
+            currentTemplate,
+            stack,
+            RequireApproval.BROADENING,
+            stack.displayName,
+            changeSet,
+          );
+          fullDiff = securityDiff.fullDiff;
+          if (securityDiff.formattedDiff) {
+            await ioHelper.notify(IO.CDK_TOOLKIT_I4400.msg(securityDiff.formattedDiff));
+            diffs += 1;
+          }
+        } else {
+          const diff = formatStackDiff(
+            ioHelper,
+            currentTemplate,
+            stack,
+            strict,
+            contextLines,
+            false,
+            stack.displayName,
+            changeSet,
+            !!resourcesToImport,
+            nestedStacks,
+          );
+
+          await ioHelper.notify(IO.CDK_TOOLKIT_I4401.msg(diff.formattedDiff), );
+          diffs += diff.numStacksWithChanges;
+          fullDiff = diff.fullDiff;
+        }
+      }
+    }
+    await ioHelper.notify(IO.CDK_TOOLKIT_I4402.msg(`Number of differences: ${diffs}`));
+
+    // Shouldn't happen but just in case
+    if (!fullDiff) {
```

Comment:
Probably the code can be changed to guarantee this

Reply:
i think this can be achieved in my subsequent PR that pulls what i can into helper methods